### ISSUE TITLE
Generating Zod schema from TS type definitions

### ISSUE THREAD

**Issue by danenania** [colinhacks/zod#53]
Love this library! I'm just beginning to convert a large TS project with many types, and I'm wondering how feasible it would be to automatically run through a file and convert any TS types to a zod schema and the accompanying inferred type. Or perhaps using the language server somehow would be easier? Either way, it would remove a ton of tedium from adopting zod for an existing project.
 
I guess I'll just barrel through and convert them one by one, which will probably take me a few hours at least, but figured I'd suggest this for those who might end up in a similar spot.

**Comment by Birowsky** [comment]
Hah! Just a few hours ago? Was just about to plead for the same feature!

But Dane, since we already have the TS types, why would we need the inferred types again?

**Comment by danenania** [comment]
@Birowsky I don't want to duplicate zod and ts definitions, so am replacing all ts types with zod schemas, then using `infer` to also get the ts type.

**Comment by Birowsky** [comment]
Oh, well then, I'll have to make a case for my kind of peoplez. I don't touch code for maybe weeks in a new project. I only model the system with types and try to find all their dependencies. Only after I feel completely comfortable with how the system is modeled, I start implementing it. So not having the types in advance, is not an option for me, which, I hope, makes a stronger case for why we would find use in Zod generator.

**Comment by vriad** [comment]
Unfortunately I have no idea how to do this sort of thing. Very open to a PR from anyone with experience with this sort of thing.

**Comment by flybayer** [comment]
I'm not sure if this is exactly what @danenania was asking for, but this would be helpful:

```ts
type Dog = {
  name: string
  neutered: boolean
}

const dogSchema = z.object<Dog>({
  name: z.string().min(3),
  neutered: z.boolean(),
});
```

And that would fail to compile if `dogSchema` did not match `Dog`

**Comment by vriad** [comment]
I think Dane was thinking more along the lines of a Babel plugin for generating Zod code, but what you just described is a _very_ powerful idea. Basically it's the inverse how how Zod works. Zod infers static types from a bunch of method calls and data structures. This is the opposite paradigm: "inferring" the internal runtime structure of a Zod schema from the TS type.

I actually already implemented this as a utility for my own purposes. I didn't put this into the Zod library because it needs TS 3.9+ to work. Here's a gist; don't panic when you see it 😛 

https://gist.github.com/vriad/074a8509cd506fdc1f96cad27cc20c77

At the bottom you can see an example of it working.

This gets *extremely* exciting when you start using it on recursive or mutually recursive types. 

```ts
type User = {
  name: string;
  age?: number | undefined;
  active: boolean | null;
  posts: Post[];
};

type Post = {
  content: string;
  author: User;
};

export const User: toZod<User> = z.late.object(() => ({
  name: z.string(),
  age: z.number().optional(),
  active: z.boolean().nullable(),
  posts: z.array(Post),
}));

export const Post: toZod<Post> = z.late.object(() => ({
  content: z.string(),
  author: User,
}));

User.shape.posts.element.shape.author;
```

The above uses a `z.late.object` method that is currently implemented but undocumented.

As far as I know this is the first time any validation library has supported a recursive object schema that still gives you access to all the methods you'd want: pick, omit, extend, partial. I've been using this to implement the API endpoint validations for a new company and it's been a dream. I'm planning to release first-party support for this in Zod 2 in a few weeks.

**Comment by flybayer** [comment]
@vriad beautiful!!! 😍 

This will be awesome with Prisma 2 since Prisma provides types all the models in your DB.

**Comment by vriad** [comment]
@flybayer  Yep, exactly :) I actually have some code lying around that auto-generates Zod schemas from a Prisma 2 client — I'll try to throw that into a separate npm module at some point. I wish the Prisma client provided a nice way of introspecting the schema but currently they don't. So I really had to dive into the guts of the client to make it work.

**Comment by flybayer** [comment]
Prisma does have undocumented introspection tools via [`@prisma/sdk`](https://github.com/prisma/prisma/blob/master/src/packages/sdk/src/index.ts). We are using them in Blitz.

But not sure if auto generating zod from the DB is worth it, since usually your input is a bit different than what's in the DB, just like you have shown in your example above.

**Comment by vriad** [comment]
I was more hoping for a way to read out a list of models/fields/relations from the Prisma client, instead of having to worry about pulling schema information down from the DB. Prisma's trying to make the .prisma file the single declarative source of truth for your schema but don't provide any easy way to build codegen tools on top of it...drives me up the wall. 

imho it still makes sense to generate Zod schemas for your base types (AKA a one-to-one mapping with what's defined in your Prisma schema). That's the ground truth, then you generate the derived types that you need from there. Plus defining the "base types" is the part that requires duplicative static/runtime declarations (because TypeScript can't infer recursive types), so if you can codegen that, you get rid of any syncing headaches.

**Comment by flybayer** [comment]
Are you in the prisma slack? I'm sure they'd love to hear what you need on the codegen front. And/or open an issue on Github.

**Comment by bradennapier** [comment]
You could use the typescript compiler api to generate using a transformer.  It's actually quite simple to do.

**Comment by vriad** [comment]
@flybayer btw I spun out the `toZod` utility from that gist into its own module. It's not in Zod core because it requires TypeScript 3.9+ (for complicated recursive type reasons).

`yarn add tozod`

```ts
import { toZod } from "tozod";

type Dog = {
  name: string
  neutered: boolean
}

const dogSchema = toZod<Dog>({
  name: z.string().min(3),
  neutered: z.boolean(),
});
```

**Comment by flybayer** [comment]
Ok thanks! I wonder if there is a way you can add it into core, but then throw an error if someone tries to use this and TS version isn't `3.9+`?

**Comment by danenania** [comment]
Interesting discussion here! We initially redefined most of our TS types as zod schemas, then used `z.infer` to extract TS types. This works pretty well, but the extra layers of inference seem to sometimes be causing performance issues with complex types, and can also break some IDE features like `Go to definition`.  So we're now looking at refactoring to use the `toZod` approach where we'll have the TS type and schema defined side by side. I originally didn't want to do this because I worried about types and schemas getting out of sync, but `toZod` seems to deal with that nicely.

**Comment by vriad** [comment]
@danenania I should warn that `toZod` is limited in what you can express. It doesn't support unions (except the special case of `.undefined` and `.nullable`, intersections, tuples, records, enums, or literals. I could add literal and enum support if you need it (I think). In general toZod is pretty fragile since its a generic recursive alias (which also can lead to other performance issues with the TS server).

If you could provide a reproducible example of the "Go to definition" issue I can look into it. I haven't encountered that.

**Comment by vriad** [comment]
@flybayer I'm not aware of a way to do that. If `toZod` is in core, then TS compilation will fail for any project <typescript@3.9. Don't think there's a way to disable the type checker based on version. :/

**Comment by bradennapier** [comment]
Export toZod separately and have them import directly?  At least could just be a choice then.

**Comment by vriad** [comment]
Unfortunately most users consume zod with `import * as z from "zod"` so there's no way to export it from `index.ts` without everyone accidentally importing it and messing up their builds.

It would be nice to enable something like `import { toZod } from 'zod/tozod';` but I'm not sure how to do that. @bradennapier Is that what you mean by "export separately"?

**Comment by bradennapier** [comment]
Yes, exactly.  It'd be quite easy.

**Comment by danenania** [comment]
Thanks for the heads up @vriad on toZod's limitations. In that case, we'll hold off on that refactoring.

This is getting pretty far from the original issue, but based on @bradennapier's demonstration that it isn't too hard to translate between TS and zod definitions, I'm starting to think the **ultimate** workflow here might be for zod to automatically generate schemas for all defined types (probably based on included files in tsconfig), and then allow for extending the auto-generated schemas to add more validations if needed. I'm sure this opens up some other cans of worms, but would make using zod a complete no-brainer in any TS project imo, since currently the only real drawback is needing to define types in zod's DSL instead of in plain TS.

**Comment by maneetgoyal** [comment]
> @danenania I should warn that `toZod` is limited in what you can express. It doesn't support unions (except the special case of `.undefined` and `.nullable`), intersections, tuples, records, enums, or literals. I could add literal and enum support if you need it (I think). In general toZod is pretty fragile since its a generic recursive alias (which also can lead to other performance issues with the TS server).
> 
> If you could provide a reproducible example of the "Go to definition" issue I can look into it. I haven't encountered that.

@vriad Do these limitations look solvable in the near future or do you think they are just technically infeasible? Trying to assess if I should be waiting on `toZod`.

Apart from the 3 (or more) use cases described above, I have a slightly tangential (maybe?) requirement which `toZod` could solve. Basically, some portions of my back-end are in Python and needs data validation. So what I've been doing is auto-converting TS definitions to JSON Schema (for doing validation with say, [this](https://python-jsonschema.readthedocs.io/en/stable/#jsonschema) library). 

After looking at Zod (which looks great for JS based validation), I tried `Zod Object Schema --zod.infer--> TS Types --typescript-json-schema--> JSON Schema`. This chain works fine but I am getting a JSON schema output which is slightly inferior in quality. 

If I define TS interfaces first, then I can end up with better JSON schema (with slightly more advanced validation). It is because `typescript-json-schema` uses the comments we write alongside the type interface declarations to improve the JSON schema output. With `zod.infer`, there is no "comment" that I can add to help `typescript-json-schema`, and hence the validation takes a slight hit.

Now if I have TS interfaces first, then the issue is to keep it in sync with Zod object schema over time as other folks mentioned. So I am kinda split between whether to maintaing both Zod object schema and TS interfaces or move the server side data validation part from Python to JS (and ditch JSON schema) or just be content with whatever JSON schema I am currently getting.

**Comment by vriad** [comment]
@maneetgoyal @danenania The design goal for toZod is to have a tool that can express database models. My goals were to express the "GraphQL grammar" (or perhaps "Prisma grammar" is a more familiar term!).

1. you can define models
2. those models can have primitive fields: strings, numbers, booleans, optionals of each, and arrays of each
3. models have relations to other models (either to-one or to-many)

Perhaps I should have named it `zod-model` to better indicate this. In this sense I don't consider tozod "incomplete". You can still use toZod to implement certain "core" types, then hook those core types together with intersections, etc.   

Can you elaborate on why you don't want to define your types in Zod and use z.infer? Dane, I'd like to hear more about your issues with "Jump to definition" and maybe see a minimal repro. 

Maneet, I think the ideal flow would be to define your JSON Schema types and codegen both Zod validators (for your JS environments) and schemas for whatever Python validation library you choose. Of course, you'd have to implement the JSON-schema => Zod codegen step. If you did, I'd happy to bring it into core and maintain it moving forward. Something like this would make it quite easy, not sure if this works or not though: https://www.npmjs.com/package/json-schema-visitor

**Comment by derekparsons718** [comment]
I love this idea that @flybayer suggested, which was partially implemented in the `tozod` library. ❤️ 

I'm looking to implement Zod in my project, but it seems like I need to strip out pretty much all of my current type definitions and replace them with Zod schemas. That effectively puts Zod in charge of my type system, which I don't like; in my opinion, Typescript should be the master of typing, and Zod should be the master of validation. One way to do that would be to make the schemas conform to my existing type system in the way flybayer suggested.

`tozod` was a great step in the right direction, but it isn't sufficient as it is. It either needs to be built out a lot more or we need another solution. Is this something that can be pursued?

To put it a different way, Zod is advertised as "Typescript first", but right now it feels more like "Zod first with Typescript as a close second". I say that because, currently, you have to write the Zod schemas first, then use them to generate types (not doing it this way leads to messy situations). To be truly "Typescript first", the Typescript types have to be king, and the schemas should conform to the types instead of (or in addition to) the types being generated from the schemas.

Would it be helpful to open this as a separate issue? I know it is not really what the author requested and is technically off topic 😜

**Comment by bradennapier** [comment]
> I love this idea that @flybayer suggested, which was partially implemented in the `tozod` library. ❤️
> 
> I'm looking to implement Zod in my project, but it seems like I need to strip out pretty much all of my current type definitions and replace them with Zod schemas. That effectively puts Zod in charge of my type system, which I don't like; in my opinion, Typescript should be the master of typing, and Zod should be the master of validation. One way to do that would be to make the schemas conform to my existing type system in the way flybayer suggested.
> 
> `tozod` was a great step in the right direction, but it isn't sufficient as it is. It either needs to be built out a lot more or we need another solution. Is this something that can be pursued?
> 
> To put it a different way, Zod is advertised as "Typescript first", but right now it feels more like "Zod first with Typescript as a close second". I say that because, currently, you have to write the Zod schemas first, then use them to generate types (not doing it this way leads to messy situations). To be truly "Typescript first", the Typescript types have to be king, and the schemas should conform to the types instead of (or in addition to) the types being generated from the schemas.
> 
> Would it be helpful to open this as a separate issue? I know it is not really what the author requested and is technically off topic 😜

I mean that is pretty much what the project i had started does precisely, @derekparsons718

**Comment by derekparsons718** [comment]
@bradennapier I love your project, it seems really helpful. However, from what I can see, it doesn't address exactly what I want. It looks like your project replaces Typescript types with equivalent Zod schemas, which is awesome and much more in line with what the original author requested; but I, on the other hand, want to keep all my types as they are and create schemas that conform to them. I do not want to *replace* my existing type definitions with schemas, I want to create schemas that *conform* to my existing types. My whole point is that I want to keep my type system in typescript and only use Zod to validate that objects match up with my existing types. (I'm sure your code could be very easily adjusted to do what I want *if* something like an advanced `tozod` was available)

I realize that we can't get away from using schemas because of the whole typescript-doesn't-exist-at-runtime thing, and that isn't really what I'm asking for. I'll give an example.

Right now, if I want to validate that an object matches `A`, I will need to *replace* my existing code...

```typescript
export interface A {
   readonly ID: number;
   delayEnd: number;
   userID: number;
   reason: string;
   taskID: number;
   initiationDate: number;
   days?: number;
   userName?: string;
}
```

...with something like this...

```typescript
//hopefully this is correct, I am still new to Zod
export const aSchema = z.object({
   delayID: z.number(),
   delayEnd: z.number(),
   userID: z.number(),
   reason: z.string(),
   taskID: z.number(),
   initiationDate: z.number(),
   days: z.number().optional(),
   userName: z.string().optional()
});

export type A = z.infer<typeof aSchema>;
```

But what I really want is something more along the lines of what `tozod` does, so instead of replacing my type definitions with a schema I can have both. Something like this:

```typescript
export interface A {
   readonly ID: number;
   delayEnd: number;
   userID: number;
   reason: string;
   taskID: number;
   initiationDate: number;
   days?: number;
   userName?: string;
}

//The use of toZod ensures that the schema matches the interface
export const aSchema: toZod<A> = z.object({
   delayID: z.number(),
   delayEnd: z.number(),
   userID: z.number(),
   reason: z.string(),
   taskID: z.number(),
   initiationDate: z.number(),
   days: z.number().optional(),
   userName: z.string().optional()
});

//Note: The above code actually would work just fine with the current `tozod` library,
//but more complex examples would easily break. Hence my original comment.
```

This preserves my original types *and* has a schema that conforms to those types. This gives me the same strong typing as using `infer<>` would have, but it leaves Typescript in charge of defining my type system and still gives me all the benefits of runtime validation. I think this scenario gives the best of all worlds and is truly "Typescript-first". I realize that it is slightly more verbose than just having a schema, but I think the benefits are well worth it. It fits much better into the established Typescript paradigm, in my opinion. I can give more extensive reasoning if desired.

Am I the only one thinking along these lines? I'm not even 100% sure that such a system is feasible, which is why I asked earlier if this is something we can even look at pursuing.

**Comment by fabien0102** [comment]
Hi everybody,
I finally did have some time to work on this crazy idea last week, and, I think this is ready to get some testers and feedbacks 😁 

I'm glad to present the very early version of [ts-to-zod](https://github.com/fabien0102/ts-to-zod), this is working with `zod@next` only, I let you try & play yourself and I'm waiting for your feedbacks 😃 

Have fun folks! 🤘 

Pro tips: Since I'm planning to use this in production, you can generate some "types integration tests" with the `--tests`, to make sure the generated schema are 100% compatible with the original types 😁 

PS: Huge thanks to @bradennapier for his POC  https://github.com/bradennapier/zod-web-converter this was a very nice starting point ❤️ and @colinhacks to have quickly implement the `.required()` (https://github.com/colinhacks/zod/issues/357)

**Comment by maxfi** [comment]
That looks awesome @fabien0102! 🎉   Can't wait to test it out. Also hoping [blitz update to zod v3](https://github.com/blitz-js/blitz/issues/2055) soon to be able to use this in blitz apps!

**Comment by derekparsons718** [comment]
Since it seems like there are at least a few other people interested in the possibility of typechecking schemas against existing types, I've opened a new issue to discuss it: #372

**Comment by flybayer** [comment]
@maxfi you can update to Zod v3 with Blitz. Blitz doesn't bundle zod, so you own the version in your package json :)

**Comment by maxfi** [comment]
Thanks @flybayer. The problem is if I update to `zod@3.0.0-alpha.4` it breaks [query/mutation resolvers](https://blitzjs.com/docs/query-resolvers). `zod@beta` is working though. 👍

**Comment by RonHouben** [comment]
You can also do simply this with just TypeScript:
```
const zodSchema = z.object({
	id: z.string(),
	title: z.string().min(5),
})

type ZodSchemaType = typeof zodSchema._output;
```

<img width="483" alt="image" src="https://user-images.githubusercontent.com/6568632/211195975-265a7b19-c1e6-4db4-a0b9-b94168f25f99.png">

**Comment by purp1eeeee** [comment]
> これが正確かどうかはわかりません@danenaniaを求めていましたが、これは非常に役立ちます!:
> 
> type Dog = {
>   name: string
>   neutered: boolean
> }
> 
> const dogSchema = z.object<Dog>({
>   name: z.string().min(3),
>   neutered: z.boolean(),
> });
> dogSchemaそして、一致しない場合、コンパイルに失敗しますDog

This fails in the current zod. ( "zod": "3.20.2")
We need to override the type in some way.
Here is my example.

![image](https://user-images.githubusercontent.com/54629159/213079296-098eb7d1-c131-4781-bee8-199f3a02b2bd.png)

**Comment by lloydjatkinson** [comment]
Is this built into Zod now or in the future? Would be great to have it!

**Comment by Roamin** [comment]
Find a online website: [https://transform.tools/typescript-to-zod](https://transform.tools/typescript-to-zod)

**Comment by ctsstc** [comment]
It seems the gist is down. Was there any movement to get this baked into Zod? I know we have `z.infer` but it would be great to have the opposite to create a zode schema from a TS Type or at least couple it and validate it against a TS type.

**Comment by rattrayalex** [comment]
The gist is now here: https://gist.github.com/colinhacks/074a8509cd506fdc1f96cad27cc20c77 and the related npm package here: https://www.npmjs.com/package/tozod but it's fairly limited and doesn't seem much used.

**Comment by IliyanID** [comment]
I know this is old but if you're looking for another simple type to define you could do this

```
export type TypeToZod<T> = {
    [K in keyof T]: undefined extends T[K] ? z.ZodOptional<z.ZodType<Exclude<T[K], undefined>>> : z.ZodType<T[K]>
}
```

Then use it like this
```
import { z } from 'zod'

type A = {
    b:string
}
const properties:TypeToZod<A> = {
b: z.string()
}

const schema = z.object(properties)

```

**Comment by ctsstc** [comment]
@IliyanID it's nice that it binds the two together, but would still be nice if it could auto-magically generate the zod type from the TS type.

**Comment by ldelia** [comment]
> 

It works like a charm! Thanks!

**Comment by nachten** [comment]
Hi There,

Like the toZod implementation but what I would like to see is a bit more of magic (don't know if it's possible, therefore i'm asking)

I would like to see something like this:

`type User = {
    id: string;
    name: string;
    age?: number | undefined;
    active: boolean | null;
    posts: Post[];
};

type Post = {
    content: string;
    author: User;
};

export const UserSchema: toZod<User> = z.late.object((model) => ({
    ...model,
    id: z.string().uuid(), // refinements are fine
}));

export const PostSchema: toZod<Post> = z.late.object((model) => ({
    ...model
}));`

Would this be possible ?

**Comment by ahmad-ali14** [comment]
Well, this is a super complex problem unless *all* of your types are set in *schema* files somehow. I believe the typescript API may be helpful in reading source files and then generating the Zod schema(s); but, it is no easy task.

**Comment by nachten** [comment]
Should I create a feature request for this ? Or can we re-open the this ticket ?

**Comment by ctsstc** [comment]
This sounds like it would require code generation, which some people are fine with and others may hate.

Folks would need to determine how the generation works. Like does it utilize a glob pattern and generate any type it finds into a zod schema? Would it utilize decorators which would then currently limit it to Typescript? Does it come with a watcher tool that regenerates on changes?

**Comment by RivasCVA** [comment]
> I know this is old but if you're looking for another simple type to define you could do this
> 
> ```
> export type TypeToZod<T> = {
>     [K in keyof T]: T[K] extends (string | number | boolean | null | undefined)
>       ? (undefined extends T[K] ? z.ZodOptional<z.ZodType<Exclude<T[K], undefined>>> : z.ZodType<T[K]>)
>       : z.ZodObject<TypeToZod<T[K]>>
> };
> ```
> 
> ...

An extension to this that fit my use case was something along the lines of:

```ts
export type TypeToZod<T> = Required<{
    [K in keyof T]: T[K] extends string | number | boolean | null | undefined
        ? undefined extends T[K]
            ? z.ZodDefault<z.ZodType<Exclude<T[K], undefined>>>
            : z.ZodType<T[K]>
        : z.ZodObject<TypeToZod<T[K]>>;
}>;
```

This will force you to give a `z.default()` value to optional properties, so when you go to `z.parse()` you do not need to worry about undefined properties.

So I can do:

```ts
const schema = createZodObject<OpenAIMessagePostRequest>({
    prompt: z.string(),
    size: z.number().default(512),
});
```

Then do:

```ts
const { prompt, size } = schema.parse(body);
```

And `size` will be of type `number` and not `undefined`.

**Comment by djmisterjon** [comment]
> An extension to @IliyanID comment above is this alternative approach that makes optional fields required with the `z.ZodDefault<>` type:
> 
> ```ts
> export type TypeToZod<T> = Required<{
>     [K in keyof T]: T[K] extends string | number | boolean | null | undefined
>         ? undefined extends T[K]
>             ? z.ZodDefault<z.ZodType<Exclude<T[K], undefined>>>
>             : z.ZodType<T[K]>
>         : z.ZodObject<TypeToZod<T[K]>>;
> }>;
> 
> export const createZodObject = <T>(obj: TypeToZod<T>) => {
>     return z.object(obj);
> };
> ```
> 
> This forces you to give a `z.default()` value to optional properties, so when you go to `parse()` the schema, you do not need to worry about undefined properties.
> 
> For example:
> 
> Given the type...
> 
> ```ts
> type Body = {
>     prompt: string;
>     size?: number;
> };
> ```
> 
> I can create the schema...
> 
> ```ts
> const schema = createZodObject<Body>({
>     prompt: z.string(),
>     size: z.number().default(512),
> });
> ```
> 
> Then do...
> 
> ```ts
> const { prompt, size } = schema.parse(body);
> ```
> 
> And `size` will be of type `number` and not `undefined`.

when trigger intellisence, this ts code freeze vscode IDE in infinity loop and crash after 2 min !
![image](https://github.com/colinhacks/zod/assets/24865815/7e445ee7-ea66-4fb5-9ead-06723672c1c5)

**Comment by PepCarmona** [comment]
Just to add for completion, if your typescript declared interface has optional fields you will want to remove that optionality:
```typescript
export type TypeToZod<T> = {
  [K in keyof T]-?: T[K] extends
    | Date
    | string
    | number
    | boolean
    | null
    | undefined
    ? undefined extends T[K]
      ? ZodOptional<ZodType<Exclude<T[K], undefined>>>
      : ZodType<T[K]>
    : ZodObject<TypeToZod<T[K]>>;
};
```

> Might want to add `Date` in the mix?
> 
> ` [K in keyof T]: T[K] extends Date | boolean | number | string | null | undefined`
> 
> > An extension to @IliyanID comment above is this alternative approach that makes optional fields required with the `z.ZodDefault<>` type:
> > ```ts
> > export type TypeToZod<T> = Required<{
> >     [K in keyof T]: T[K] extends string | number | boolean | null | undefined
> >         ? undefined extends T[K]
> >             ? z.ZodDefault<z.ZodType<Exclude<T[K], undefined>>>
> >             : z.ZodType<T[K]>
> >         : z.ZodObject<TypeToZod<T[K]>>;
> > }>;
> > 
> > export const createZodObject = <T>(obj: TypeToZod<T>) => {
> >     return z.object(obj);
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This forces you to give a `z.default()` value to optional properties, so when you go to `parse()` the schema, you do not need to worry about undefined properties.
> > For example:
> > Given the type...
> > ```ts
> > type Body = {
> >     prompt: string;
> >     size?: number;
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I can create the schema...
> > ```ts
> > const schema = createZodObject<Body>({
> >     prompt: z.string(),
> >     size: z.number().default(512),
> > });
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then do...
> > ```ts
> > const { prompt, size } = schema.parse(body);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And `size` will be of type `number` and not `undefined`.

**Comment by Harm-Nullix** [comment]
I think what you mean is, if you would want to remove the optional fields, you can use this approach?  

Because removing optional fields, in my view, feels a bit weird because they _can_ be there. They should by typed optionally with their type or undefined, so `T | undefined` to say it simply.

> Just to add for completion, if your typescript declared interface has optional fields you will want to remove that optionality:
> 
> ```ts
> export type TypeToZod<T> = {
>   [K in keyof T]-?: T[K] extends
>     | Date
>     | string
>     | number
>     | boolean
>     | null
>     | undefined
>     ? undefined extends T[K]
>       ? ZodOptional<ZodType<Exclude<T[K], undefined>>>
>       : ZodType<T[K]>
>     : ZodObject<TypeToZod<T[K]>>;
> };
> ```
> 
> > Might want to add `Date` in the mix?
> > ` [K in keyof T]: T[K] extends Date | boolean | number | string | null | undefined`
> > > An extension to @IliyanID comment above is this alternative approach that makes optional fields required with the `z.ZodDefault<>` type:
> > > ```ts
> > > export type TypeToZod<T> = Required<{
> > >     [K in keyof T]: T[K] extends string | number | boolean | null | undefined
> > >         ? undefined extends T[K]
> > >             ? z.ZodDefault<z.ZodType<Exclude<T[K], undefined>>>
> > >             : z.ZodType<T[K]>
> > >         : z.ZodObject<TypeToZod<T[K]>>;
> > > }>;
> > > 
> > > export const createZodObject = <T>(obj: TypeToZod<T>) => {
> > >     return z.object(obj);
> > > };
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > This forces you to give a `z.default()` value to optional properties, so when you go to `parse()` the schema, you do not need to worry about undefined properties.
> > > For example:
> > > Given the type...
> > > ```ts
> > > type Body = {
> > >     prompt: string;
> > >     size?: number;
> > > };
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I can create the schema...
> > > ```ts
> > > const schema = createZodObject<Body>({
> > >     prompt: z.string(),
> > >     size: z.number().default(512),
> > > });
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Then do...
> > > ```ts
> > > const { prompt, size } = schema.parse(body);
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > And `size` will be of type `number` and not `undefined`.

**Comment by cunj123** [comment]
Just to add a small improvement to @IliyanID's answer, here's `TypeToZod` which allows arrays as property types

```
export type TypeToZod<T> = {
    [K in keyof T]: T[K] extends (infer E)[]
        ? z.ZodArray<z.ZodType<E>>
        : T[K] extends string | number | boolean | null | undefined
          ? undefined extends T[K]
              ? z.ZodOptional<z.ZodType<Exclude<T[K], undefined>>>
              : z.ZodType<T[K]>
          : z.ZodObject<TypeToZod<T[K]>>;
};
```

**Comment by mallanisp** [comment]
> > An extension to [@IliyanID](https://github.com/IliyanID) comment above is this alternative approach that makes optional fields required with the `z.ZodDefault<>` type:
> > export type TypeToZod<T> = Required<{
> >     [K in keyof T]: T[K] extends string | number | boolean | null | undefined
> >         ? undefined extends T[K]
> >             ? z.ZodDefault<z.ZodType<Exclude<T[K], undefined>>>
> >             : z.ZodType<T[K]>
> >         : z.ZodObject<TypeToZod<T[K]>>;
> > }>;
> > 
> > export const createZodObject = <T>(obj: TypeToZod<T>) => {
> >     return z.object(obj);
> > };
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This forces you to give a `z.default()` value to optional properties, so when you go to `parse()` the schema, you do not need to worry about undefined properties.
> > For example:
> > Given the type...
> > type Body = {
> >     prompt: string;
> >     size?: number;
> > };
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I can create the schema...
> > const schema = createZodObject<Body>({
> >     prompt: z.string(),
> >     size: z.number().default(512),
> > });
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then do...
> > const { prompt, size } = schema.parse(body);
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And `size` will be of type `number` and not `undefined`.
> 
> when trigger intellisence, this ts code freeze vscode IDE in infinity loop and crash after 2 min ! ![image](https://github.com/colinhacks/zod/assets/24865815/7e445ee7-ea66-4fb5-9ead-06723672c1c5)

Add "| Array<any>" and you're golden!

**Comment by francisduvivier** [comment]
Very surprised that nobody here suggested just using z.infer and type assignment like shown below for the type-schema comparison.
Seems a lot simpler and robust than the solutions suggested above.

```
// Helper type
type CheckSame<
  TS_TYPE,
  TS_TYPE_REPEAT extends INFERRED_ZOD_TYPE,
  INFERRED_ZOD_TYPE extends TS_TYPE,
> = never;


// Usage
type Dog = {
  name: string;
  neutered: boolean;
};

const dogSchema = z.object({
  name: z.string().min(3),
  neutered: z.boolean(),
});

type Checks = [CheckSame<Dog, Dog, z.infer<typeof dogSchema>>];
```

**Edit: this approach does not work for optional properties. Solution for that [commented here](https://github.com/colinhacks/zod/issues/53#issuecomment-3194163878)**

**Comment by yuzheng14** [comment]
Maybe you can see my project: https://github.com/yuzheng14/valype , it converts typescript definitions to validators directly!

Suppose you have typescript definition:

```ts
export interface User {
  name: string
  age: number
}
```

Valype can generate code like:

```ts
import { z } from 'zod/v4'
import { $ZodIssue } from 'zod/v4/core'

export interface User {
  name: string
  age: number
}

const UserSchema = z.object({
  name: z.string()
  age: z.number()
}

export function validateUser(data: unknown): $ZodIssue[] | undefined {
  return UserSchema.safeParse(data).error?.issues
}
```

Additionally, valype has unplugin so it can be integrated in Vite, Rollup, Webpack, etc., seamlessly.

The official site is https://valype.dev

I'll appreciate if you give me some feedback to make Valype better!

**Comment by swelborn** [comment]
> Very surprised that nobody here suggested just using z.infer and type assignment like shown below for the type-schema comparison. Seems a lot simpler and robust than the solutions suggested above.
> 
> ```
> // Helper type
> type CheckSame<
>   TS_TYPE,
>   TS_TYPE_REPEAT extends INFERRED_ZOD_TYPE,
>   INFERRED_ZOD_TYPE extends TS_TYPE,
> > = never;
> 
> 
> // Usage
> type Dog = {
>   name: string;
>   neutered: boolean;
> };
> 
> const dogSchema = z.object({
>   name: z.string().min(3),
>   neutered: z.boolean(),
> });
> 
> type Checks = [CheckSame<Dog, Dog, z.infer<typeof dogSchema>>];
> ```

https://github.com/colinhacks/zod/issues/2807#issuecomment-1977406654

Cross-posting this comment that seems to work...

**Comment by francisduvivier** [comment]
> > Very surprised that nobody here suggested just using z.infer and type assignment like shown below for the type-schema comparison. Seems a lot simpler and robust than the solutions suggested above.
> > ```
> > // Helper type
> > type CheckSame<
> >   TS_TYPE,
> >   TS_TYPE_REPEAT extends INFERRED_ZOD_TYPE,
> >   INFERRED_ZOD_TYPE extends TS_TYPE,
> > > = never;
> > 
> > 
> > // Usage
> > type Dog = {
> >   name: string;
> >   neutered: boolean;
> > };
> > 
> > const dogSchema = z.object({
> >   name: z.string().min(3),
> >   neutered: z.boolean(),
> > });
> > 
> > type Checks = [CheckSame<Dog, Dog, z.infer<typeof dogSchema>>];
> > ```
> 
> [#2807 (comment)](https://github.com/colinhacks/zod/issues/2807#issuecomment-1977406654)
> 
> Cross-posting this comment that seems to work...

That is a nice construct, and thanks for the playground link! However, there are 2 issues: 

- extra properties in the zod schema will not trigger a TS error
- if there is an optional property in the TS Type which is absent in the zod type, it will not trigger a typescript error
For covering these issues, I've eventually ended with this construct:

```
type EqualTypes<X, Y> =
  (<T>() => T extends X ? 1 : 2) extends <T>() => T extends Y ? 1 : 2
    ? true
    : false;

export const __tsCheckSame = <T1, T1_REPEAT extends T2, T2 extends T1>(
  value: EqualTypes<T1, T2>
) => undefined;

// Usage
__tsCheckSame<Dog, Dog, z.infer<typeof dogSchema>>(true);
```

Advantage here is that you get a clear explanation for missing non-optional properties, and for the missing optional properties, you still get an error, even though the error will not say what exactly is wrong.
More details of usage in [this playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcBEyEAJgPQBuAzPgNwBQdMAnmAKZwCiAjgK4CGAGwAqLVgGcAPAA0ANHACaAPjgBeOnDgAKCUMWaAlKuVC4rAB4xWAO2Ji4UuAH44ARjgAuOACZD5yzbsdPUMVY1MLa1sFJ1cPb3UNGJgoHlYEjU9MQTFWejpzSFg4AGMIKzF4AH1KmDEAYQALVmKAawBlPhB2FTgdFzkhF0qAJQ4ABQ4AQRM-SLshLwGvcP8owb0E8kFUz25+YVFJQaXFOhDlHhtWTGArVmI8hmY2ODaeACMAeXeAK2b4HoIBJiD5jHBsWBMTwVKC3ADm9A0EDAMGAZUEb3eYORrEhjmhyXh9DQeWe7AAIhA4apEAkrJ1WATYVYEXTWDxLFB7p53hAIAJWHwrIi4MjUeiBNiIcx8XAYUTgR9vn9ijBPJjlf9iXkBfAQV9fv82sUmiA+DTkAA6CCG1WaIEafVS3HMTxW+UsgwyBJitH0gSY52Qt2Wj1wgzWlF+wReuhofT0UrleDEKnG03mnpWm0qmD2ukMkNhiMgW6aSj6b0aO4c3HclCW3n8wVWWNIqMSoOuhvF-SR8X+ttypW2tXDg259OsM3e+N5aq1RrNdoMiSUuFyddyK23TC4iRkiCYOCpuFTs2KPTJVIJhikUhwAACtQAtAV-m+oDgoHBS2IxPCcBWGUL6+hKcBgOCLrAOIcB8FyJ6sJYqr3HAADuwAwA0iFiMUsJRuQ7C4t+dALvUTStB0XQSAgaByLR26Wru+6Hsep7nnwl6aNerC3owohwOuACywD-vCnwdv6NKfKWMBrlScj4GB-pdkw+CKPONTkcuVGsBICTCaJAEshJA6CFWglUiJYkmZJ5kJDuVh7lAB6iEeJ5pia06cXQGwaPeT6vu+qqft+v5GYBymCBBUGwDBdjwewxBIf8qF8HY+A8SUQrAfA7zsOlAFwncxBwDAEBYNkrD4AkPFnHkSYVB5Z5eWa1nGXCpnRgINLsa1fDWrJ9qinZkqxd2WVzqRWlLpRq5bpZG4NsxLmsc1HHteJo1cQkAXPmIb5mGwIXEdA4U2dSUU9ZBOJxbBiWIchlilelBBZcUOUQHlBVicVqHlZVAg5DVGh1XxZGzSu1EGVZEW2WZAgWYZF1dRKFmOc5rlsO5fUZpt8PdacV4pLxpICcjHUaqNNIOkBhZyoSLIijWnL1k2ApCiKV2qbKYYivqmqqp4tOOqC41QgzzKshoJJxg1ZRNbj3n42eSrU1m1qjvm1b0+6jPhn2pathWFks3WxAhuzLZDtz4tFvrEZXUOAujiGOb-NriROnbPYO5WCTxrO4MzRRUN6TDcIq1TCNI7DF3R916NMU5LFuWxnkZkTmi7Q++2HcdMChWdf4dSNCMxbdqL3QhyVPWlGXvZ931wb9JVlRVWRA9VtUk-VDAQ6HunyUtC0Y6n2Ppy1eNw6rXzbX5cB7UFR0fqdP4l5F1M3dK8VwTXKUoS9Dck9lVi5XA+Ut0VbcA53wM96kffTYug+rhHUdqzH78zwnaMOcnmM1pKzaj-T+hNfLcRJreIAA)

**Comment by francisduvivier** [comment]
I've landed on this approach for now:

```
type EqualTypes<X, Y> =
  (<T>() => T extends X ? 1 : 2) extends <T>() => T extends Y ? 1 : 2
    ? true
    : false;

export const __tsCheckSame = <T1, T1_REPEAT extends T2, T2 extends T1>(
  value: EqualTypes<T1, T2>
) => undefined;

// Usage
__tsCheckSame<Dog, Dog, z.infer<typeof dogSchema>>(true);
```

It's not perfect still, but the advantage here is that you get a clear explanation for missing non-optional properties.
For the missing optional properties, you still get an error, even though the error will not say what exactly is wrong.
More details of usage in [this playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcBEyEAJgPQBuAzPgNwBQdMAnmAKZwCiAjgK4CGAGwAqLVgGcAPAA0ANHACaAPjgBeOnDgAKCUMWaAlKuVC4rAB4xWAO2Ji4UuAH44ARjgAuOACZD5yzbsdPUMVY1MLa1sFJ1cPb3UNGJgoHlYEjU9MQTFWejpzSFg4AGMIKzF4AH1KmDEAYQALVmKAawBlPhB2FTgdFzkhF0qAJQ4ABQ4AQRM-SLshLwGvcP8owb0E8kFUz25+YVFJQaXFOhDlHhtWTGArVmI8hmY2ODaeACMAeXeAK2b4HoIBJiD5jHBsWBMTwVKC3ADm9A0EDAMGAZUEb3eYORrEhjmhyXh9DQeWe7AAIhA4apEAkrJ1WATYVYEXTWDxLFB7p53hAIAJWHwrIi4MjUeiBNiIcx8XAYUTgR9vn9ijBPJjlf9iXkBfAQV9fv82sUmiA+DTkAA6CCG1WaIEafVS3HMTxW+UsgwyBJitH0gSY52Qt2Wj1wgzWlF+wReuhofT0UrleDEKnG03mnpWm0qmD2ukMkNhiMgW6aSj6b0aO4c3HclCW3n8wVWWNIqMSoOuhvF-SR8X+ttypW2tXDg259OsM3e+N5aq1RrNdoMiSUuFyddyK23TC4iRkiCYOCpuFTs2KPTJVIJhikUhwAACtQAtAV-m+oDgoHBS2IxPCcBWGUL6+hKcBgOCLrAOIcB8FyJ6sJYqr3HAADuwAwA0iFiMUsJRuQ7C4t+dALvUTStB0XQSAgaByLR26Wru+6Hsep7nnwl6aNerC3owohwOuACywD-vCnwdv6NKfKWMBrlScj4GB-pdkw+CKPONTkcuVGsBICTCaJAEshJA6CFWglUiJYkmZJ5kJDuVh7lAB6iEeJ5pia06cXQGwaPeT6vu+qqft+v5GYBymCBBUGwDBdjwewxBIf8qF8HY+A8SUQrAfA7zsOlAFwncxBwDAEBYNkrD4AkPFnHkSYVB5Z5eWa1nGXCpnRgINLsa1fDWrJ9qinZkqxd2WVzqRWlLpRq5bpZG4NsxLmsc1HHteJo1cQkAXPmIb5mGwIXEdA4U2dSUU9ZBOJxbBiWIchlilelBBZcUOUQHlBVicVqHlZVAg5DVGh1XxZGzSu1EGVZEW2WZAgWYZF1dRKFmOc5rlsO5fUZpt8PdacV4pLxpICcjHUaqNNIOkBhZyoSLIijWnL1k2ApCiKV2qbKYYivqmqqp4tOOqC41QgzzKshoJJxg1ZRNbj3n42eSrU1m1qjvm1b0+6jPhn2pathWFks3WxAhuzLZDtz4tFvrEZXUOAujiGOb-NriROnbPYO5WCTxrO4MzRRUN6TDcIq1TCNI7DF3R916NMU5LFuWxnkZkTmi7Q++2HcdMChWdf4dSNCMxbdqL3QhyVPWlGXvZ931wb9JVlRVWRA9VtUk-VDAQ6HunyUtC0Y6n2Ppy1eNw6rXzbX5cB7UFR0fqdP4l5F1M3dK8VwTXKUoS9Dck9lVi5XA+Ut0VbcA53wM96kffTYug+rhHUdqzH78zwnaMOcnmM1pKzaj-T+hNfLcRJreIAA)

**Comment by Harm-Nullix** [comment]
#This thread should not be used anymore

Zod 4 now has the z.input function, which does this inhouse:
[https://zod.dev/basics#inferring-types](https://zod.dev/basics#inferring-types)
```
const mySchema = z.string().transform((val) => val.length);
 
type MySchemaIn = z.input<typeof mySchema>;
// => string
 
type MySchemaOut = z.output<typeof mySchema>; // equivalent to z.infer<typeof mySchema>
// number
```

imo

**Comment by kwilcz** [comment]
It's not perfect, since it forces you to use zed to hold your schema, when most of the times you need only validation.

Also, it was discussed here already, see https://github.com/colinhacks/zod/issues/53#issuecomment-664672096

**Comment by vdawg-git** [comment]
And sometimes you dont control the type but need to use it.

So z.infer is not a solution
